### DISCUSSION TITLE
Inverter that is offline

### DISCUSSION THREAD

**Discussion by klaasnicolaas** [klaasnicolaas/python-omnikinverter#24] [General]
A current problem is that most inverters switch off when there is no energy production from the sun. This causes the problem that every request that follows results in an exception error, resulting in a log file (in Home Assistant) that gets flooded with exceptions.

How would a python package best deal with this case:
1. Do not print an exception within a certain time range.
2. Print another exception after sunset, just once.
3. etc.

**Comment by freddieleeman** [comment]
If the last known power value within 24 hours was 0, no errors at all.
If the inverter power drops from > 0W to 'unavailable' write error
If the inverter has been 'unavailable' for over 24 hours write error

**Reply by freddieleeman** [reply]
Caching would also be necessary while unavailable:

- Power stays 0
- Total energy should be cached indefinitely
- daily cache until 00h00
- month until last day of month 00h00
- year until last day of year 00h00

**Comment by robbinjanssen** [comment]
Another option is to "cache" the value until 00:00.

For example; 
- 17:00 -> 18 kWh today, current power 1200 watt, 2000 kWh total
- 18:00 -> 19 kWh, current power 1100 watt, 2001 kWh total
- 19:00 -> Exception (return 19 kWh, 0 watt, 2002 kWh total)
- ....
- 00:00 -> Exception (return 0 kWh, 0 watt, 2002 kWh total)

**Comment by klaasnicolaas** [comment]
It is not the intention that we will cache for month and year totals, for that we have the energy dashboard long-term statistics. I'm now purely concerned with the error handling, something that unfortunately cannot really be prevented I think.

The behavior we want to prevent is that an entity for total just jumps to 0 kWh, which causes an energy spike in the dashboard.

**Reply by freddieleeman** [reply]
Agreed, I personally only use the Power and Lifetime Energy values. Both can be cached when the device in unavailable as they do not change when power is 0.

**Comment by sander76** [comment]
My inverter (zeversolar) can fail in two ways:

1. It is off. No sun, but that might also occur during the day.
2. There is an unkown response which is useless to me. A second request results in correct values. This raises a parsing error.

For now focussing on 1, I think the below pseudo code should work.

```python
try:
    return get_inverter_data()
except TimeoutError:
    if timeout_exceeds_24_hrs():  # as @freddieleeman suggest
        raise IntegrationError()
    if new_day_has_arrived():
        return {"daily_energy_use": 0, "current_solar_power": 0}
        # don't update "lifetime energy values
        # setting "current_solar_power" will reset the sensor 
        # Provided it is defined with state class total increasing as defined in 
        # https://developers.home-assistant.io/docs/core/entity/sensor#long-term-statistics
    return {"current_solar_power": 0}
        # don't update "daily_energy_use" and "lifetime energy values"
```

**Comment by hepfister** [comment]
My two cents. I'm not a coder and will explain in a functional way. 
I use this 'no cloud' plugin and the other Omnik plugin simultaneously, originally to see which one works best for me. No cloud caches the Total value (until midnight). The other plugin just drops to zero. For both I declared a device_class and both work fine with the Energy Dashboard. Choosing either one seems fine from an Energy Dashboard perspective it seems. However, the nicest option in my opinion would be that the Energy total is cached until a new value is received. So return to last known value after the inverter switches off and until the inverter gives data the next day. Wouldn't this not fix the exception / flooded logs problem?

**Comment by sander76** [comment]
This is what I currently have running on my zeversolar inverter dealing with the same challenge.

https://github.com/sander76/home-assistant/tree/zever_local/homeassistant/components/zeversolar

It uses the caching and the new day detection mentioned by @freddieleeman and @hepfister . It's running fine on my instance for some days now. No flooding of the logs anymore.

See below my daily energy generation.

![chart](https://user-images.githubusercontent.com/1688712/133216770-33b66103-a301-471d-b948-1fa41c7c0589.png)

**Comment by klaasnicolaas** [comment]
I think we should move this discussion towards the [integration side](https://github.com/robbinjanssen/home-assistant-omnik-inverter/tree/dev) of Omnik and let's see how we can integrate some form of caching (no longer via a file) like zeversolar. maybe @sander76  can you help us with what adjustments are needed on the dev branch?

**Comment by sander76** [comment]
Happy to help. 

I guess it all revolves around the `_async_update_data` method defined in your repo [here](https://github.com/robbinjanssen/home-assistant-omnik-inverter/blob/868fb1f13a723d00fc01e127bf8c4463dc8ca914/custom_components/omnik_inverter/__init__.py#L92)
and [mine](https://github.com/sander76/home-assistant/blob/zever_local/homeassistant/components/zeversolar/__init__.py#L79)

The "caching" taking place is looking at the last successful solar data (which is assumed to be always valid) stored in the `data` attribute of the update coordinator.
look [here](https://github.com/sander76/home-assistant/blob/605fd0f0317429a51e5727fcb2c27d8115db78d7/homeassistant/components/zeversolar/__init__.py#L100) and [here](https://github.com/sander76/home-assistant/blob/605fd0f0317429a51e5727fcb2c27d8115db78d7/homeassistant/components/zeversolar/__init__.py#L104)

I would also only focus on daily energy and current power. total energy is tracked by hass itself.

**Reply by hepfister** [reply]
> I would also only focus on daily energy and current power. total energy is tracked by hass itself.

Are you sure this is the case?I thought the reverse is true. The Energy dashboard only needs Total Energy and Hass derives Daily Energy from its own calculations. Daily Energy = current Total Energy value - start Total Energy value form that day (>0.00h). The Current Power value isn't used add all. This is how I have setup my Energy Dashboard, using only Total Energy.

**Reply by sander76** [reply]
I guess that would depend how you define the `state_class` of the device. https://developers.home-assistant.io/docs/core/entity/sensor#available-state-classes

I have set mine as `total_increasing` and the reset cycle is each day. 
If you would set it as `measurement` I think you can use the "total_energy" property of the inverter.

**Reply by sander76** [reply]
you're right, the power value isn't used. But I found it an interesting measurement as it gives me a kind of realtime insight.
